sysboundary/inflow.cpp

```diff
@@ -0,0 +1,497 @@
+      creal dz = cellParams[CellParams::DZ];
+      creal x = cellParams[CellParams::XCRD] + 0.5 * dx;
+      creal y = cellParams[CellParams::YCRD] + 0.5 * dy;
+      creal z = cellParams[CellParams::ZCRD] + 0.5 * dz;
+
+      isThisCellOnAFace.fill(false);
+      determineFace(isThisCellOnAFace.data(), x, y, z, dx, dy, dz);
+      // Comparison of the array defining which faces to use and the array telling on which faces this cell is
+      doAssign = false;
+      for (int j = 0; j < 6; j++) {
+         doAssign = doAssign || (facesToProcess[j] && isThisCellOnAFace[j]);
+      }
+      if (doAssign) {
+         mpiGrid[id]->sysBoundaryFlag = this->getIndex();
+      }
+   }
+
+   // Assign boundary flags to local fsgrid cells
+   const std::array<int, 3> gridDims(technicalGrid.getLocalSize());
+   for (int k = 0; k < gridDims[2]; k++) {
+      for (int j = 0; j < gridDims[1]; j++) {
+         for (int i = 0; i < gridDims[0]; i++) {
+            const auto coords = technicalGrid.getPhysicalCoords(i, j, k);
+
+            // Shift to the center of the fsgrid cell
+            auto cellCenterCoords = coords;
+            cellCenterCoords[0] += 0.5 * technicalGrid.DX;
+            cellCenterCoords[1] += 0.5 * technicalGrid.DY;
+            cellCenterCoords[2] += 0.5 * technicalGrid.DZ;
+            const auto refLvl = mpiGrid.get_refinement_level(mpiGrid.get_existing_cell(cellCenterCoords));
+
+            if (refLvl == -1) {
+               abort_mpi("Error, could not get refinement level of remote DCCRG cell!", 1);
+            }
+
+            creal dx = P::dx_ini * pow(2, -refLvl);
+            creal dy = P::dy_ini * pow(2, -refLvl);
+            creal dz = P::dz_ini * pow(2, -refLvl);
+
+            isThisCellOnAFace.fill(false);
+            doAssign = false;
+
+            determineFace(isThisCellOnAFace.data(), cellCenterCoords[0], cellCenterCoords[1], cellCenterCoords[2],
+                          dx, dy, dz);
+            for (int iface = 0; iface < 6; iface++) {
+               doAssign = doAssign || (facesToProcess[iface] && isThisCellOnAFace[iface]);
+            }
+            if (doAssign) {
+               technicalGrid.get(i, j, k)->sysBoundaryFlag = this->getIndex();
+            }
+         }
+      }
+   }
+}
+
+void Inflow::applyInitialState(const dccrg::Dccrg<SpatialCell, dccrg::Cartesian_Geometry>& mpiGrid,
+                               FsGrid<std::array<Real, fsgrids::bfield::N_BFIELD>, 2>& perBGrid, Project& project) {
+   for (uint popID = 0; popID < getObjectWrapper().particleSpecies.size(); ++popID) {
+      setCellsFromTemplate(mpiGrid, popID);
+   }
+   setBFromTemplate(mpiGrid, perBGrid);
+}
+
+void Inflow::updateState(const dccrg::Dccrg<SpatialCell, dccrg::Cartesian_Geometry>& mpiGrid,
+                         FsGrid<std::array<Real, fsgrids::bfield::N_BFIELD>, 2>& perBGrid, creal t) {
+   if (t - tLastApply < tInterval) {
+      return;
+   } else {
+      tLastApply = t;
+   }
+#pragma omp parallel for
```

Comment:
I would assume that threading will be more beneficial deeper in when looping over cells or over v-blocks. Although possibly I am the one who put a parallel for in such initialisation loops a decade ago. But that will require some care if e.g. moments are computed there.

Reply:
Still I would say remove this and thread deeper in.